### PR TITLE
Feat: Add skill book reset. Fix skill book open/save. Fix mob stats.

### PR DESCRIPTION
**Pull request by quotopian**
## Additions
- Skill book can now reset allocations.
- Skills can now consume SP and EP (stamina).

## Fixes
- Mob stats are now parsed correctly.
- Skill book now saves skill points correctly.
- Skill book now opens correctly with a functioning UI.
---------------------------------------------------------------------
TODO: Add multiple skill tabs.

### CODE REVIEW

File: GameDataParser/Parsers/SkillParser.cs [hunk 1]
```diff
@@ -21,34 +21,33 @@ protected override List<SkillMetadata> Parse()
...
-                    SkillMetadata metadata = new SkillMetadata();
-                    List<SkillLevel> skillLevels = new List<SkillLevel>();
-                    metadata.SkillId = int.Parse(skillId);
                     XmlDocument document = Resources.XmlMemFile.GetDocument(entry.FileHeader);
+                    XmlNodeList levels = document.SelectNodes("/ms2/level");
 
-                    metadata.Type = byte.Parse(document.SelectSingleNode("/ms2/basic/kinds").Attributes["type"].Value);
+                    string skillId = Path.GetFileNameWithoutExtension(entry.Name);
+                    string skillType = document.SelectSingleNode("/ms2/basic/kinds").Attributes["type"].Value;
+                    bool skillRecovery = (int.Parse(document.SelectSingleNode("/ms2/basic/kinds").Attributes["spRecoverySkill"]?.Value ?? "0") > 0) ? true : false;
```
Comment (angelotadeucci) [inline GameDataParser/Parsers/SkillParser.cs:29]:
```suggestion
                    bool skillRecovery = int.Parse(document.SelectSingleNode("/ms2/basic/kinds").Attributes["spRecoverySkill"]?.Value ?? "0") > 0;
```

File: Maple2Storage/Types/NpcStats.cs [hunk 2]
```diff
@@ -81,29 +81,37 @@ public struct NpcStats
...
         }
+
+        public NpcStat(T max)
+        {
+            Bonus = max;
+            Base = max;
+            Total = max;
+        }
+
         public NpcStat(T total, T min, T max)
```
Comment (angelotadeucci) [inline Maple2Storage/Types/NpcStats.cs:110]: Should rename the parameters too so it's less confusing

File: MapleServer2/PacketHandlers/Game/JobHandler.cs [hunk 3]
```diff
@@ -18,12 +21,17 @@ public override void Handle(GameSession session, PacketReader packet)
...
             {
-                case 8: // Close Skill Tree
+                case 0x08: // Close Skill Tree
                     HandleCloseSkillTree(session);
                     break;
-                case 9: // Save Skill Tree
+                case 0x09: // Save Skill Tree
                     HandleSaveSkillTree(session, packet);
                     break;
+                case 0x0A:
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/JobHandler.cs:30]: Should use an enum for the cases

File: MapleServer2/PacketHandlers/Game/SkillBookTreeHandler.cs [hunk 4]
```diff
@@ -17,12 +17,17 @@ public override void Handle(GameSession session, PacketReader packet)
...
             {
-                case 0: // Open skill tree
+                case 0x00: // Open skill tree
                     HandleOpen(session);
                     break;
-                case 1: // Save skill tree
+                case 0x01: // Save skill tree
                     HandleSave(session);
                     break;
+                case 0x04:
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/SkillBookTreeHandler.cs:26]: Use enum here too

File: MapleServer2/Packets/JobPacket.cs [hunk 5]
```diff
@@ -35,41 +35,72 @@ public static Packet Save(Player character, int objectId = 0)
...
             {
                 if (id == countId)
                 {
                     pWriter.WriteByte(split); // Write that there are (split) skills left
                 }
                 pWriter.WriteByte();
-                pWriter.WriteByte(skills[id].Learned);
-                pWriter.WriteInt(id);
-                pWriter.WriteInt(skills[id].SkillLevels.Select(x => x.Level).FirstOrDefault());
+                pWriter.WriteByte((skills[id] > 0) ? 1 : 0);    // Is it learned?
```
Comment (angelotadeucci) [inline MapleServer2/Packets/JobPacket.cs:63]:
```suggestion
                pWriter.WriteBool(skills[id] > 0);    // Is it learned?
```

**Review by angelotadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/JobHandler.cs [hunk 6]
```diff
@@ -18,12 +21,17 @@ public override void Handle(GameSession session, PacketReader packet)
...
+                    HandleResetSkillTree(session, packet);
+                    break;
+                default:
+                    break;
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/JobHandler.cs:34]: Add default logger like other handlers

File: MapleServer2/Packets/JobPacket.cs [hunk 7]
```diff
@@ -35,41 +35,72 @@ public static Packet Save(Player character, int objectId = 0)
...
+                pWriter.WriteInt(Math.Clamp(skills[id], skillData[id].SkillLevels.Select(x => x.Level).FirstOrDefault(), int.MaxValue));    // Level to display
                 pWriter.WriteByte();
             }
             pWriter.WriteShort(); // Ends with zero short
 
             return pWriter;
         }
 
+        // KMS2
+        //public static Packet WriteSkills(this PacketWriter pWriter, Player character)
```
Comment (AlanMorel) [inline MapleServer2/Packets/JobPacket.cs:74]: Why?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Packets/JobPacket.cs (hunk 7 above)
Comment (quotopian) [reply MapleServer2/Packets/JobPacket.cs:74]: I was planning to remove out of draft.

**Review by angelotadeucci — APPROVED** [review]

**Review by AlanMorel — APPROVED** [review]

**Pull request merged**